### DISCUSSION TITLE
Invisible Text Removal

### DISCUSSION THREAD

**Discussion by Raks-coder** [pymupdf/PyMuPDF#905] [Q&A] [answered]
Hi, 
Will it be possible to add a feature similar to adobe acrobat pro dc to remove hidden text specifically the text in white colour which is hidden in white background?
There are a few PDFs which have invisible texts which become visible when the PDF is selected in document viewer.

**Comment by JorjMcKie** [comment, marked as answer]
This is a complex undertaking, unfortunately. Because when we know the text color, we do not know, where in the page definition source (i.e. the `/Contents`) this text is being written - which we must in order to delete it.
Another issue (however solvable) is finding out whether the text background is white.
Of course we could use redaction annotations to remove text, but this incurs the risk of removing other stuff unintendedly.
I have also been asked to remove text, that is covered by other objects like images: same issue here - I would be forced not only to check what comes first: image or text, but also whether the image is transparent (in which case the text is not hidden), etc.

**Comment by HelloKitty2022** [comment]
hi is any way of removing some text that font size >50 @ @JorjMcKie

**Comment by HelloKitty2022** [comment]
sometimes these text overlap with nomal text,

**Comment by HelloKitty2022** [comment]
replace or removing text is really usefull  
but mupdf not 
is technically difficult?or concerned about copyright?

**Comment by JorjMcKie** [comment]
- [x] is technically difficult
    * It is a pain in the neck and error prone - you hinted at some difficutlies yourself (overlaps with other stuff). Whatever you get to work there, you will also find examples that won't. This is due to PDF's complexity - its dozens of fundamentally different ways to achieve same-looking results: your example is one out of many ways how to "hide" text. There is a nice and well-defined way: text rendering mode 3. But no, _viva la peculiaridad!_
- [ ] or concerned about copyright
    * No, there are no license issues with MuPDF whatsoever - it is free software, as long as you don't create a commercial product with it. Same is true for PyMuPDF by the way. It has ever been like that implicitely, because it doesn't work without MuPDF. Now this is also stated explictely.

If you want, look at the source code behind `document.scrub`, where text is removed that is hidden by using rendering mode 3. Try finding a way to extend this ...

**Comment by HelloKitty2022** [comment]
thanks a lot . really brilliant lib

**Comment by JorjMcKie** [comment]
> hi is any way of removing some text that font size >50

Sorry, I didn't read your question closely enough: this is actually doable without using redactions, because fontsize is part of the text "object" inside the page definition. In principle  - and **_after_** you did a `page.clean_contents()` - you will see the following lines when a text object id defined:

```
...
BT
/F1 12 Tf
<text output commands>
/F2 50 Tf    %
<text output commands> %
/F3 20 Tf
<text output commands>
ET
...
```
Each text object is wrapped in string pair b"BT" / b"ET". Start using some font is a line ending with string b"Tf". The number to the left (a float) is the font size (12, 50, 20 above).
This is valid until a new Tf or end of text object (ET).
So you **_could_** simpley delete the lines above marked with an "%" to get rid of text with an undesired fonntsize. In printciple the following snippet:
```python
page.clean_contents()
xref = page.get_contents()[0]  # xref of the /Contents
cont = bytearray(doc.xref_stream(xref))  # read the contents source modifyable
contlines = cont.splitlines()
# locate a line having `.endswith(b"Tf")`. Assume line number = i
# extract fontsize
# if fontsize < 50 continue searching for Tf
# else look for next line that either ends with b"Tf" or is equal to b"ET". Line number = j
del contlines[i, j] # remove thos lines
# when done with all removals:
doc.update_stream(xref, b"\n".contlines)  # update / write back contents
```

**Comment by mani2106** [comment]
[This](https://github.com/pymupdf/PyMuPDF/issues/736) is a related conversation which can help @Raks-coder
